### DISCUSSION TITLE
Add binary on flathub

### DISCUSSION THREAD

**Discussion by roukmoute** [rustdesk/rustdesk#1537] [General]
I've seen there is a flatpak ready.

Is it possible to add it on [flathub](https://github.com/flathub/flathub/wiki/App-Submission)?

Thanks in advance

**Comment by Root-Core** [comment]
Just to channel the discussion so far I will add some quotes.

[Quote](https://github.com/rustdesk/rustdesk/issues/459#issuecomment-1245127847) @rustdesk:

> The Flatpak version can not be used as controlled side, the container (sandbox) has big limitation to us.

[Quote](https://github.com/rustdesk/rustdesk/pull/1492#issuecomment-1245122386) @rustdesk:

> So flatpak can not be used as controlled side.

---------------

I think there is a major misconception, mistake or miscommunication.
Flatpack does not allow the implementation of daemons, while Snap does.

@rustdesk thinks this (or the sandboxing) makes the implementation of the ability to control the system running the Flatpack version impossible.

----------------

On the other hand, there is [AnyDesk](https://flathub.org/apps/details/com.anydesk.Anydesk) on Flathub, which just does this.
They have implemented the "daemon" into a tray icon, which stays in backgrounds and listens for incoming requests.
This seems to be a viable solution for this problem and to me like the preferred implementation anyways.

----------------

At this point it is not clear what exactly hinders the implementation or how AnyDesk archives it.

**Reply by My1** [reply]
I might wanna link the flathub anydesk repo which likely contains all the stuff they used to make anydesk into a flatpack from the outside
https://github.com/flathub/com.anydesk.Anydesk

**Comment by sandr01d** [comment]
Thanks for the summary @Root-Core. I think there are a few things I can add to this regarding what is still missing for flatpak support. 

The [manifest](https://github.com/sandr01d/flathub/blob/rustdesk/com.rustdesk.rustdesk.yaml) is ready. When building for flatpak, [inline-sciter.py](https://github.com/rustdesk/rustdesk/blob/master/res/inline-sciter.py) need to be run and the build needs to be made with `--feature inline,flathub`, so we'd need an extra flatpak release on the release page.

Currently rustdesk fails to determine the correct display server when run in a flatpak, because `get_display_server_of_session()` relies on loginctl, which isn't available in flatpak.
https://github.com/rustdesk/rustdesk/blob/551c5ecddfd6d5a7adff261f36592fece2874d27/libs/hbb_common/src/platform/linux.rs#L9
I currently don't know how to solve this one, so if anyone has an idea, please let me know.

Regarding the system tray: This seems to be wanted for Linux anyway, since #809 has been added to #918. The [crate](https://github.com/open-trade/trayicon-rs) that is currently used for a tray icon only supports Windows, but maybe the maintainers would be fine with replacing it with a crate that is multi platform? Also #912 could probably tackled with this one as well.

That being said, I personally think the flatpak could be shipped before system tray support on Linux is implemented. I can see some use cases where being able to run rustdesk in the background is not required.

**Reply by istori1** [reply]
`flatpak-spawn --host loginctl`

For that to work this permission is needed
`- --talk-name=org.freedesktop.Flatpak`

**Reply by rustdesk** [reply]
@Kingtous

**Reply by sandr01d** [reply]
> `flatpak-spawn --host loginctl`
> 
> For that to work this permission is needed `- --talk-name=org.freedesktop.Flatpak`

Thanks for the input @istori1! I've created #1731 to implement this. Are you sure `- --talk-name=org.freedesktop.Flatpak` is needed? I'm asking because it seems to work fine without it.

**Comment by My1** [comment]
as a SteamDeck user I would also greatly think this as an awesome idea. not sure how exactly anydesk works but it certainly works, and an open source relay-compatible application that can run on multiplatform is something way too rare imo

**Reply by roukmoute** [reply]
> as a SteamDeck user I would also greatly think this as an awesome idea. not sure how exactly anydesk works but it certainly works, and an open source relay-compatible application that can run on multiplatform is something way too rare imo

It is currently not possible for me to use rustdesk, but if you would like to use another, you can use NoMachine, I explain how here: https://www.reddit.com/r/SteamDeck/comments/x9pe5y/comment/iq33k76/?utm_source=share&utm_medium=web2x&context=3

**Reply by My1** [reply]
or I can use anydesk for now which not only plain works as a flatpak, but crucially doesnt need to disable the os-readonly which (notably) protects partitions that are axed upon update. I dont wanna get weird problems because stuff has gone missing after the update and it is needed for something and all that.

**Reply by istori1** [reply]
Also in Flatpak:
https://github.com/LizardByte/Sunshine

https://github.com/LizardByte/Sunshine/releases/download/v0.14.1/sunshine.flatpak

https://github.com/istori1/Easy-Tan#easy-tan-sunshine-quick-setup-for-linux

**Comment by ghost** [comment]
> I've seen there is a flatpak ready.
> 
> Is it possible to add it on [flathub](https://github.com/flathub/flathub/wiki/App-Submission)?
> 
> Thanks in advance

Waiting on this, too. The current beta 1.2.0 has flatpak, though it isn't submitted on Flathub

**Reply by FredsterCode** [reply]
can you share a link to that beta please. I would try it..

**Reply by sandr01d** [reply]
Can be found on the releases page as nightly:

https://github.com/rustdesk/rustdesk/releases

**Reply by p-fruck** [reply]
@rustdesk is there any update on this? From https://github.com/rustdesk/rustdesk/issues/3973 is seems like the flatpak will be working fine on multiple systems including the steam deck with the next release. Are there any considerations to publish the flatpak to some repository like flathub.org?

**Reply by tesfabpel** [reply]
As a user I believe it's better to have it on flathub (maybe marked as prerelease if possible...) and then fix the various issues later if the flatpak version is mostly+ working than not having it there at all. It's also better for Steam Deck users since the system is by default RO.

**Reply by My1** [reply]
> since the system is by default RO.

not just that but enabling write can lead to more problems than it solves due to A/B Partitioning and likely image based updates.

**Reply by realitymolder** [reply]
Version 1.2.0 is out but no flatpak on flathub.

**Comment by tim77** [comment]
Precompiled binary [not enough](https://reproducible-builds.org/). Example of how to properly build canonical from sources Rust projects on Flathub:
- https://github.com/flathub/net.veloren.veloren
- https://github.com/flatpak/flatpak-builder-tools/tree/master/cargo

**Comment by fatmalama** [comment]
what is keeping it out from flathub now?

**Reply by rustdesk** [reply]
Contributors

**Reply by Justinzobel** [reply]
Would you be OK with me adding it? I'm happy to help maintain it as I already help maintain others for developers, including the whole KDE stack.

**Reply by sandr01d** [reply]
Would also be interested to know what needs to be done. Maybe I can give a helping hand too. I created a [manifest](https://github.com/sandr01d/flathub/tree/rustdesk) a while back. I guess it would need some changes though since this was before the transition from sciter to flutter. My code changes to make rustdesk work in a flatpak have been merged though, so I think we should be good on that side.

**Reply by Root-Core** [reply]
Some time ago, I created a manifest that automatically fetches and outputs a flatpak file from the official Debian package.
It executes and opens up a window, but the window is empty and does not work.

I also added an AppStream file.. which might be useful independent of the flatpak.
I will try to find and clean up my work, maybe it is useful to get this finally solved.

**Reply by realitymolder** [reply]
@Root-Core that would be amazing to see as a repo or program, it can help to compline many open source projects to flatpaks. Especially today in the immutable OS era!

**Reply by FurtadoPires** [reply]
May I suggest to hold this until `1.2.4` release as stable? Because this version it's the first one that supports wayland remote control, and since release on flathub might be the first contact a lot of people will have with this project, would be better if it works out of the box for wayland users.

**Reply by rustdesk** [reply]
Thanks @Justinzobel @sandr01d @Root-Core

**Reply by rustdesk** [reply]
> Would you be OK with me adding it?

Please go ahead, @Justinzobel

**Reply by rustdesk** [reply]
> flutter

@sandr01d yes, and adding arm64

**Reply by LichiMan** [reply]
Hi! I'm very happy Rustdesk will finally arrive to Flathub.

Will the Flathub version be able to autostart the server at boot someway even without the needs to login first?
Is this possible?

Thanks.

**Reply by xlionjuan** [reply]
@LichiMan This is not how Flatpak works

**Comment by rustdesk** [comment]
https://github.com/flathub/flathub/pull/5233

**Reply by FurtadoPires** [reply]
Which version the branch `master` is current using? Because I almost sure they'll require the package to be built directly from it. You may need to start to work with develop branches in this project.

I personally suggest the [git flow](https://www.gitkraken.com/learn/git/git-flow) approach in case it becomes necessary :)

**Reply by Croydon** [reply]
Why would they require a certain branch? git tags represent versions

**Reply by FurtadoPires** [reply]
Probably because it's considered to be a good practice making master / main branch == latest stable release

It's common for users who prefer to compile / builder softwares / [AUR](https://aur.archlinux.org/) maintainers to pull directly to the main branch of the repository, so developers usually put any new / not tested / incomplete code in other branches like beta / release candidate / etc.

Judging by how this repository is right now, I can guess that any modifications are being included directly to the master branch, which can result in a different build from the release version.

That's why I suggested the use of git flow (linked above) or another branch strategy that rustdesk developers feel comfortable to work with, because my guess is that flathub curators won't like the package not being build directly from master.

**Reply by Croydon** [reply]
Sorry, but I think your points are completely off-topic regarding putting Rustdesk on Flathub. Flatpaks on Flathub use most of the time specific stable versions and if not [then some specific commit](https://github.com/flathub/org.fedoraproject.MediaWriter/blob/bcd23216c915c7cdd32fbe0fcf212133db13f678/org.fedoraproject.MediaWriter.json#L30). Where the source downloads come from is irrelevant as long as they are official and stay unchanged, i.e. match the saved checksums.

Rustdesk's Flathub should stick to stable release versions as it also does with other downloads / installer / package formats. And what the work flow is in this main repository here does not matter for Flathub.

**Reply by FurtadoPires** [reply]
> Sorry, but I think your points are completely off-topic regarding putting Rustdesk on Flathub.

So I check up the request, and one of the curators asked

> why not building from source?

I know that's because it's being built using the deb file from `1.2.4` (which is not currently tagged as the stable) instead of directly from the source code.

To be clear, I'm not saying that's **obligatory** to do it, I'm saying that is considered a good practice and they may not like it (the way it's being doing right now). But if they accept without any changes in that part I mentioned then great, it means that it hopefully ready to be approved and maintained and my suggestion is pointless (as you stated).

**Reply by nalsai** [reply]
Flatpak does not require it to be built from source but you are supposed to do it if you can. 

But what I believe Croydon was referring to is that there are no special changes needed for that.
It is enough to specify the tag [`1.2.4`](https://github.com/rustdesk/rustdesk/tree/1.2.4) or the commit 9827c7651442800f47428087903d52da8cb8c2c8.

**Reply by joshua-stone** [reply]
@rustdesk Hey, thank you for trying to submit Rustdesk to Flathub! I had attempted to build it from source with the goal of eventually submitting it to Flathub, however I kept getting blocked whenever trying to generate build sources via [`flatpak-cargo-generator`](https://github.com/flatpak/flatpak-builder-tools/tree/master/cargo):

```
$ ./flatpak-cargo-generator.py Cargo.lock
INFO:root:Loading packages from https://github.com/21pages/android-wakelock
INFO:root:Adding package android-wakelock from https://github.com/21pages/android-wakelock
INFO:root:Loading packages from https://github.com/fufesou/arboard
From https://github.com/fufesou/arboard
 * branch            956b5f8693b4fc7fddd7b8cafbe1111a892b34b1 -> FETCH_HEAD
Warning: you are leaving 2 commits behind, not connected to
any of your branches:

  9e6f535 cargo fmt
  f520dba Use msecs instead of duration

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 9e6f535

HEAD is now at 956b5f8 Set custom timeout for x11 server connection.
INFO:root:Adding package arboard from https://github.com/fufesou/arboard
INFO:root:Loading packages from https://github.com/clslaid/cacao
INFO:root:Loading packages from https://github.com/rustdesk-org/confy
INFO:root:Adding package confy from https://github.com/rustdesk-org/confy
INFO:root:Loading packages from https://github.com/madsmtm/core-foundation-rs
INFO:root:Adding package core-foundation from https://github.com/madsmtm/core-foundation-rs
INFO:root:Adding package core-foundation-sys from https://github.com/madsmtm/core-foundation-rs
INFO:root:Adding package core-graphics from https://github.com/madsmtm/core-foundation-rs
INFO:root:Adding package core-graphics-types from https://github.com/madsmtm/core-foundation-rs
INFO:root:Loading packages from https://github.com/fufesou/evdev
INFO:root:Adding package evdev from https://github.com/fufesou/evdev
INFO:root:Loading packages from https://github.com/21pages/hwcodec
From https://github.com/21pages/hwcodec
 * branch            1d504ee590c15472813fecc22cee4b8149b2b8cd -> FETCH_HEAD
Fetching submodule deps
From https://github.com/21pages/deps
 + 940b80b...718280b android    -> origin/android  (forced update)
Warning: you are leaving 3 commits behind, not connected to
any of your branches:

  0973290 0.4.16
  315d8d0 windows signature
  2894e34 update example

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 0973290

HEAD is now at 1d504ee 0.4.15
INFO:root:Adding package hwcodec from https://github.com/21pages/hwcodec
INFO:root:Loading packages from https://github.com/21pages/impersonate-system
INFO:root:Adding package impersonate_system from https://github.com/21pages/impersonate-system
INFO:root:Loading packages from https://github.com/rustdesk-org/keepawake-rs
From https://github.com/rustdesk-org/keepawake-rs
 * branch            ad94454a75cf1ff9e95e217dee9dd6a378bf625e -> FETCH_HEAD
Warning: you are leaving 2 commits behind, not connected to
any of your branches:

  64d5685 Merge pull request #3 from Nadahar/apple-sys
  b19a985 - Replaced apple-sys with a minimal IOKit binding - Fixed VSCode settings

If you want to keep them by creating a new branch, this may be a good time
to do so with:

 git branch <new-branch-name> 64d5685

HEAD is now at ad94454 Merge pull request #2 from 21pages/set_display
INFO:root:Adding package keepawake from https://github.com/rustdesk-org/keepawake-rs
INFO:root:Loading packages from https://github.com/21pages/machine-uid
INFO:root:Adding package machine-uid from https://github.com/21pages/machine-uid
INFO:root:Loading packages from https://github.com/rustdesk-org/magnum-opus
INFO:root:Adding package magnum-opus from https://github.com/rustdesk-org/magnum-opus
INFO:root:Loading packages from https://github.com/fufesou/mouce
INFO:root:Adding package mouce from https://github.com/fufesou/mouce
INFO:root:Loading packages from https://github.com/fufesou/pam
INFO:root:Adding package pam from https://github.com/fufesou/pam
INFO:root:Loading packages from https://github.com/rustdesk-org/parity-tokio-ipc
INFO:root:Adding package parity-tokio-ipc from https://github.com/rustdesk-org/parity-tokio-ipc
INFO:root:Loading packages from https://github.com/fufesou/rdev
INFO:root:Adding package rdev from https://github.com/fufesou/rdev
INFO:root:Loading packages from https://github.com/rustdesk-org/reqwest
INFO:root:Adding package reqwest from https://github.com/rustdesk-org/reqwest
INFO:root:Loading packages from https://github.com/open-trade/pulsectl
INFO:root:Adding package rust-pulsectl from https://github.com/open-trade/pulsectl
INFO:root:Loading packages from https://github.com/open-trade/rust-sciter
INFO:root:Adding package sciter-rs from https://github.com/open-trade/rust-sciter
INFO:root:Loading packages from https://github.com/rustdesk-org/sysinfo
INFO:root:Adding package sysinfo from https://github.com/rustdesk-org/sysinfo
INFO:root:Loading packages from https://github.com/rustdesk-org/tao
INFO:root:Adding package tao from https://github.com/rustdesk-org/tao
INFO:root:Adding package tao-macros from https://github.com/rustdesk-org/tao
INFO:root:Loading packages from https://github.com/fufesou/the-fat-controller
INFO:root:Adding package tfc from https://github.com/fufesou/the-fat-controller
INFO:root:Loading packages from https://github.com/rustdesk-org/tokio-socks
INFO:root:Adding package tokio-socks from https://github.com/rustdesk-org/tokio-socks
INFO:root:Loading packages from https://github.com/tauri-apps/tray-icon
From https://github.com/tauri-apps/tray-icon
 * branch            b8dbd42c6f94a29f34b0a0daa619486277185512 -> FETCH_HEAD
Previous HEAD position was 9231438 ci: recognize contributors in releases
HEAD is now at b8dbd42 Apply Version Updates From Current Changes (#117)
INFO:root:Adding package tray-icon from https://github.com/tauri-apps/tray-icon
INFO:root:Loading packages from https://github.com/21pages/wallpaper.rs
INFO:root:Adding package wallpaper from https://github.com/21pages/wallpaper.rs
INFO:root:Loading packages from https://github.com/21pages/rust-webm
INFO:root:Adding package webm from https://github.com/21pages/rust-webm
INFO:root:Adding package webm-sys from https://github.com/21pages/rust-webm
INFO:root:Loading packages from https://github.com/bjornsnoen/x11-rs
INFO:root:Adding package x11 from https://github.com/bjornsnoen/x11-rs
INFO:root:Loading packages from https://github.com/clslaid/x11-clipboard
INFO:root:Adding package x11-clipboard from https://github.com/clslaid/x11-clipboard
Traceback (most recent call last):
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 436, in <module>
    main()
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 429, in main
    generated_sources = asyncio.run(generate_sources(load_toml(args.cargo_lock),
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib64/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 383, in generate_sources
    for pkg in await asyncio.gather(*pkg_coros):
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 333, in get_package_sources
    return await get_git_package_sources(package, git_repos)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 270, in get_git_package_sources
    git_repo['commits'][commit] = await get_git_repo_packages(repo_url, commit)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 206, in get_git_repo_packages
    get_cargo_toml_packages('.')
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 191, in get_cargo_toml_packages
    cargo_toml = load_toml('Cargo.toml')
                 ^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/home/jstone/Work/github/rustdesk/rustdesk/./flatpak-cargo-generator.py", line 96, in load_toml
    toml_data = toml.load(f)
                ^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/toml/decoder.py", line 156, in load
    return loads(f.read(), _dict, decoder)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/toml/decoder.py", line 471, in loads
    raise TomlDecodeError("An implicitly defined "
toml.decoder.TomlDecodeError: An implicitly defined table can't be an array (line 64 column 1 char 1903)
```

I filed a bug report here:

https://github.com/flatpak/flatpak-builder-tools/issues/384

Do you have any insight on what may be going wrong?

If building the flatpak from source isn't feasible, then using release packages as build sources should suffice.

I have prior experience with submitting apps to Flathub, and I can help do the same for Rustdesk if there's any interest.

**Reply by macmpi** [reply]
@rustdesk not sure what's the permissions issue which seemingly prevents flathub submission: can you elaborate please?

**Reply by nalsai** [reply]
The two permission problems I saw reading the submission were "--filesystem=home" and "--talk-name=org.freedesktop.Flatpak".

Regarding "--filesystem=home", it is only needed if the application needs full access to the user's home or directories. 
I haven't used rustdesk exstensively, so I don't know whether it is truly needed or not, but if you just want to be able to select and send files that can be done using portals, a feature that natively works with most toolkits and allows the user to choose a file that then gets sent to the application.
Rustdesk was not able to justify why they need full access, and their only argument was that other software has it too. Just because someone has robbed a bank before doesn't mean I should do it. (They argued similarly for other issues)
It should be a simple matter to either use portals or state the actual reason why complete access is needed. (Using an old toolkit without support for portals and not being able to update or fix it / needing access to directories to browse them (remotely) from within the UI)

Secondly, "--talk-name=org.freedesktop.Flatpak" is a major point, as it would allow breaking out of the sandbox and completely defeating security. Therefore, it is not allowed without getting an exception.
It looks like it is needed for running loginctl (https://github.com/rustdesk/rustdesk/discussions/1537), something they only wrote after saying stuff like "These features are really required for remote desktop software", "it is really needed", referring to manifests of other remote desktop software that does not need it, and "If you still do not know what is remote desktop, here is a wiki about it, en.wikipedia.org/wiki/Remote_desktop_software". If rustdesk absolutely needs it, this issue could be solved by requiring the user to manually set it if they need it (and adding it to the readme + display a message if the user tries to use it without having set the permission). In that case, the rustdesk flatpak could for example function as just a client until it is manually given the permission.


They were incapable of writing a proper manifest without lots of guidance.
Everything necessary can be found at https://docs.flatpak.org/en/latest/ and by looking at other Flathub manifests and submissions.

**Reply by My1** [reply]
For the file access thing it's quite simple. Rustdesk allows a direct file transfer where i can browse both local and remote sides and pull from and push to locations i select. Using portals would make this overly complicated i would assume especially as you would have to act on the screen at the other side to navigate the portal (which considering rustdesk iirc also supports pure file transfer connections, isn't overly feasible)

Honestly not sure what the talk-name thing does maybe it's there to be able to start a service to do unattended and notably pre-login connections.

**Reply by sandr01d** [reply]
Regarding `--talk-name=org.freedesktop.Flatpak`: RustDesk uses `loginctl` to get information about the running session (display server, session id, seat, state etc.). In the case of flatpak `loginctl` is run outside the sandbox using `flatpak-spawn --host` to be able to get this information from the host. See https://github.com/rustdesk/rustdesk/blob/37ebac2a9e01b0a4c7e17645109521bdfcdc6b7b/libs/hbb_common/src/platform/linux.rs#L237C1-L255C2. `flatpak-spawn` requires access to the org.freedesktop.Flatpak D-Bus interface. I do not know whether there are other reasons why `--talk-name=org.freedesktop.Flatpak` would be needed, but `flatpak-spawn` is not used anywhere else, so probably not.
Unfortunately I'm not aware of an alternative way to access this kind of information from within a flatpak, but I would be happy to implement it if there is any.

**Reply by Darin755** [reply]
Maybe I'm wearing a tin foil hat but to me Rustdesk seems sketchy. Little is known about the developers and Rustdesk itself uses lots of permissions and could theoretically take over a significant amount of systems. The fact that the distribution of Rustdesk comes from a single source makes me a bit concerned that a mass take over of machines running Rustdesk could happen. I want a flathub version as it would significantly lower risk on the Linux side. Once again, I'm wearing a tin hat. I just know that this could be XZ but instead of failing it could be massively successful.

**Reply by My1** [reply]
well sure I understand the worry, especially in terms of the flatpak thing I would have loved to at least bring in my point in regards to the file system perms but apparently the discussion is locked there, so I cant add it.

**Reply by jvschiavo** [reply]
> Maybe I'm wearing a tin foil hat but to me Rustdesk seems sketchy. Little is known about the developers and Rustdesk itself uses lots of permissions and could theoretically take over a significant amount of systems. The fact that the distribution of Rustdesk comes from a single source makes me a bit concerned that a mass take over of machines running Rustdesk could happen. I want a flathub version as it would significantly lower risk on the Linux side. Once again, I'm wearing a tin hat. I just know that this could be XZ but instead of failing it could be massively successful.

Couldn't the same be said about TeamViewer and AnyDesk? Actually it is even worse, since both of them are proprietary, so the users can't access the source code, and both are hugely more popular. Rustdesk not only is free software, but it is AGPLv3 licensed. So we even have access to the server side so we can host ourselves. It really adds to the credibility in my opinion.

I also have a lot of concerns about security, but these facts alone makes RustDesk much more trustworthy in my concept. But yes, being available on some reputable software repositories like Flathub and specially Debian main would give me much more peace of mind about it.

**Reply by My1** [reply]
true true about others doing the same but not being open, but the problem is that rustdesk couldnt even properly articulate why they need the permissions when asked and basically said "the others use it too" or "it's needed for this kind of app" without explaining WHY it is needed specifically.

this makes them appear less competent than they actually are.

**Reply by jvschiavo** [reply]
I think it its a miscommunication problem. The RustDesk devs don't seem to be native in English, and I guess they felt personally targeted by being treated differently than those other popular apps.

There was the other "scandal" about the ["Chinese certificate"](https://github.com/rustdesk/rustdesk/discussions/6444), which was honestly a poorly-placed, weak, test certificate for a display driver, but it had a legitimate purpose, for legitimate features (virtual display, privacy mode etc), and the only thing it was really capable of was "Code Signing" despite seeming otherwise. They went after it and bought the EV certificate (U$270.00/year), just so people could feel more calm about this.
And the Chinese characters came from an encoding problem to UTF-16 in Windows.

Anyway, I just hope this can be resolved anytime soon.

**Comment by olumolu** [comment]
Will rustdesk try this again to upload this app in flathub.
I don't understand the issues but what I understand that this makes no sense that small apps are even coming into flathub why not rustdesk.if there is a issue due to dependency's and stuff that can be fixed. Most users don't know about rustdesk as most users use apps from store and this is really important to have the app in store to make people aware and install and use.

**Reply by My1** [reply]
rustdesk should **carefully** plan out how to respond to obvious questions next time tho rather than saying something similar to "anydesk etc use those permissions too" and calling double standards. They were literally putting up an argument kids use in literal kindergarten, or "We add these features not for playing, it is because we really need. Hope you can understand".

This is one of the primary things they should "fix".

Flathub maintainers cannot know every piece of software and what it needs, so it's up to the dev to *properly* explain what the perms are being used for.

for example "why does rustdesk ask for full file system access rather than e.g. using portals?"

you could for example try to answer.
"similar to other solutions we offer a full file browser to the remote side which allows them to upload files directly where they should go and download files directly if needed from their site without any added latencies or interruptions when controlling over a low speed internet connection.
additionally we offer users the ability to connect solely for file transfer without even seeing the remote desktop to control any portals for even more efficiency"

or for the permission to use GPU acceleration and audio they could mention how they want to have
1) accelerated en/decoding of the stream
2) the ability to relay audio to the client.

and not answering (direct quote from the thread:
"This is remote desktop software, these are required to run it. Please check anydesk https://github.com/flathub/com.anydesk.Anydesk/blob/master/com.anydesk.Anydesk.json"

for the socket and talk permissions I do not know enough about flatpak to even know what these exactly do but i would guess they are in part needed for the screen recording?

**Reply by olumolu** [reply]
@rustdesk can you clarify why this is not in flathub till now.
What is stopping you from do this.
If anydesk can be in flathub why not rustdesk.
Update is hard as it is not in flathub. There is no point not to have that.
If rustdesk has some major issue if yes lets discuss.
Why this ignore behaviour from rustdesk why is the issue.
Flathub have there discuss forum the issues can be shorted out and provide the app to a larger user.
I have faced many issues i ask to use rustdesk and he could not find it in software center as it is not in flathub. I highly recommend to put this app in flathub. So users can use this without any such issue.

**Reply by My1** [reply]
basically rustdesk gave up after they were asked why permissions were needed, and rustdesk was basically just saying "anydesk uses them too, double standards", and so on.

**Reply by jvschiavo** [reply]
> If anydesk can be in flathub why not rustdesk.

That's a very good question, one that you should take to the Flathub team. Have the Anydesk devs ever explained why they use the permissions they use? And if not, would Flathub be willing to completely remove it from the repo until this is resolved? :eyes: I don't think they have the guts to do that.

Flatpak/Flathub seems to always have been about "Popular proprietary crap first, Promised security features Later™ (not in our lifetime)."

**Reply by My1** [reply]
the anydesk package is iirc not even from anydesk.

also the issue about adding anydesk was made by a flatpak member themselves apparently so there wasnt any discussion, at least in the relevant PR.

**Reply by nalsai** [reply]
Rustdesk also tried to have the permission "--talk-name=org.freedesktop.Flatpak" by default which would allow it to control flatpak and therefore break out of its container, nullifying any security mechanisms of flatpak. No other remote desktop application has that as far as I know.

Anydesk does not have any unreasonable permissions like that and rustdesk would be allowed just fine with the same permissions as anydesk, if they can justify their need for them. They just need to write a proper flatpak manifest and briefly explain why the permissions are needed.

**Reply by My1** [reply]
so that permission would basically be what flatseal uses to modify permissions of other flatpak things?
that's actually quite a big thing.

**Reply by olumolu** [reply]
Can anyone do this so we can have the app as 3rd party .

**Reply by My1** [reply]
also just in case it isnt obvious, the addition of anydesk happened unilaterally from flathub, they cant really go and ask anydesk to modify their software for flathub, considering anydesk possibly didnt even exactly consent let alone request it, but rustdesk is on its own requesting to be added and they obviously have the power to change stuff.

**Reply by My1** [reply]
Cool. As one can read in the comments, it helps not being dismissive about perms or hell, having a permission to bypass the sandbox, it literally just asks for filesystem /home, internet, x11 and ipc, and the file system permission was properly answered

**Comment by rustdesk** [comment]
You are welcome to fork https://github.com/rustdesk-org/flathub and submit to flathub.

> https://github.com/flathub/flathub/pull/5233

**Reply by besdar** [reply]
@rustdesk hello! The Fluthub repository has been forked and a new PR is open.: https://github.com/flathub/flathub/pull/5670

**Reply by My1** [reply]
Very good commenting on the PR in my opinion, especially on justifying permissions, well done.
I also added a few lines to be a bit more specific, while also staying constructive and added a question that might help with the issue of loginctl

**Reply by olumolu** [reply]
Nice to see its happening

**Reply by rustdesk** [reply]
Good job

**Comment by rustdesk** [comment]
https://github.com/flathub/com.rustdesk.RustDesk

**Reply by joshua-stone** [reply]
Excellent! Now the app just has to get verified:

https://github.com/flathub/com.rustdesk.RustDesk/issues/1

**Reply by besdar** [reply]
I agree with @joshua-stone

What do you think about it, @rustdesk?

As I see, all you need is to [open an issue](https://github.com/flathub/flathub/issues/new) with permission request to [the Flathub's repository](https://github.com/flathub/com.rustdesk.RustDesk). Then you will be able to verify the app in Flathub.

info: 
1) https://docs.flathub.org/docs/for-app-authors/verification#id-like-to-verify-my-app-that-someone-else-already-published
2) https://docs.flathub.org/docs/for-app-authors/maintenance/#maintainer-access-to-flathub-application-repository

**Reply by rustdesk** [reply]
https://github.com/flathub/flathub/issues/5677?

**Reply by besdar** [reply]
@rustdesk great! But you should ask them to add your github account as a maintainer of [com.rustdesk.RustDesk](https://github.com/flathub/com.rustdesk.RustDesk) repository. Then you will be able to [login to the flathub website](https://flathub.org/login) and see this admin panel that I now have:
![image](https://github.com/user-attachments/assets/53f80870-34ee-4d7f-ba67-aca228d66652)

And there you will be able to verify your app automatically by using your website `rustdesk.com`

**Reply by besdar** [reply]
@rustdesk they say you should already have an invitation. Please check this link: https://github.com/flathub/com.rustdesk.RustDesk/invitations

**Reply by rustdesk** [reply]
Verification is done https://github.com/flathub/flathub/issues/5677#issuecomment-2397009710

**Reply by joshua-stone** [reply]
Looks great! Thank you!

![image](https://github.com/user-attachments/assets/6221f53a-2f7f-41a5-8065-bd03dbcc7716)

**Comment by rustdesk** [comment]
https://flathub.org/apps/com.rustdesk.RustDesk

**Reply by olumolu** [reply]
This is awesome

**Reply by Aqua1ung** [reply]
> https://flathub.org/apps/com.rustdesk.RustDesk

I am guessing that's the stable version, right (i.e. _not_ the nightly)?

**Reply by besdar** [reply]
Yes, it is a stable version

**Reply by Aqua1ung** [reply]
> Yes, it is a stable version

Thanks. So hopefully you guys will continue issuing the flatpak _nightly_ in Github, right?

**Reply by xlionjuan** [reply]
> > Yes, it is a stable version
> 
> Thanks. So hopefully you guys will continue issuing the flatpak _nightly_ in Github, right?

I don't see why not, it is in the CI already

**Reply by olumolu** [reply]
@rustdesk provide a link in the download button with flathub so users can directly open flathub and install easily
